test/common/network/listener_impl_test.cc

```diff
@@ -102,4 +104,81 @@ TEST(ListenerImplTest, UseOriginalDst) {
   dispatcher.run(Event::Dispatcher::RunType::Block);
 }
 
+TEST(ListenerImplTest, FallbackToWildcardListener) {
+  Stats::IsolatedStoreImpl stats_store;
+  Event::DispatcherImpl dispatcher;
+  Network::TcpListenSocket socket("tcp://127.0.0.1:10000", true);
+  Network::TcpListenSocket socketDst("tcp://0.0.0.0:10001", false);
+  Network::MockListenerCallbacks listener_callbacks1;
+  Network::MockConnectionHandler connection_handler;
+  // The virtual listener of exact address does not exist, fall back to wild card virtual listener.
+  Network::TestListenerImpl listener(connection_handler, dispatcher, socket, listener_callbacks1,
+                                     stats_store, {.bind_to_port_ = true,
+                                                   .use_proxy_proto_ = false,
+                                                   .use_original_dst_ = true,
+                                                   .per_connection_buffer_limit_bytes_ = 0});
+  Network::MockListenerCallbacks listener_callbacks2;
+  Network::TestListenerImpl listenerDst(connection_handler, dispatcher, socketDst,
+                                        listener_callbacks2, stats_store,
+                                        Network::ListenerOptions());
+
+  Network::ClientConnectionPtr client_connection =
+      dispatcher.createClientConnection(Utility::resolveUrl("tcp://127.0.0.1:10000"));
+  client_connection->connect();
+
+  Address::InstancePtr alt_address(new Address::Ipv4Instance("127.0.0.1", 10001));
+  EXPECT_CALL(listener, getOriginalDst(_)).WillRepeatedly(Return(alt_address));
+  EXPECT_CALL(connection_handler, findListenerByAddress(alt_address))
```

Comment:
The unit test of findlistenerbyaddress is in connection_handler_test.cc. This here is for use_original_dst, which can't be tested there.